### ISSUE TITLE
INVESTIGATE: Generalized Teacher Forcing — the training-time version of blend_recurrent_state, with a testable prediction for w≈0.1

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#294]
> **Investigative issue.** Establish whether the correspondence is real and whether the mechanism
> transfers, before building anything. Dossier + pre-registration with numeric falsifiers first — and
> see #290/#291's checks for the standard: a *numeric* rule, committed before the check runs.

## We appear to have re-invented this at inference time

**M38/M39/M41** (PR #292): blending the ConvLSTM **cell** state back toward its last real-data value
during free-running recovers **+0.039 AP@h18**, the effect is entirely in the cell (not the hidden
half), and the response **saturates at w≈0.1** — a 10% pull per step already buys 83% of a hard clamp.
We implemented it as:

```python
blend_recurrent_state:  weight * anchor + (1 - weight) * new
```

**Hess et al. 2023** propose, as a *training* method (claim confidence: **proven**):

> Generalized Teacher Forcing (GTF) — linearly interpolating between RNN-generated states and target
> states as **z̃ = (1−α)·z + α·z̄** at each training step — **strictly bounds the Jacobian product series
> for arbitrary time horizons when α is chosen as α = 1 − 1/σ_max**, where σ_max is the supremum of
> spectral norms over all RNN Jacobians (**requiring σ_max ≥ 1**; below 1 the formula yields negative
> α, outside the valid range). This solves the exploding gradient problem for training on chaotic
> systems without constraining the model's ability to exhibit chaos.

**Same functional form, same axis, opposite side of the training/inference line.**

## ⭐ The falsifier that makes this worth doing, and it is quantitative

GTF does not just share our form — **it predicts the value of our knob.**

If our empirically-optimal `weight` corresponds to GTF's α, then M41's saturation at **w ≈ 0.1** implies

```
0.1 = 1 − 1/σ_max     ⇒     σ_max ≈ 1.11
```

**σ_max is measurable on weights we already have**, with no training: the supremum of the spectral norm
of the ConvLSTM Jacobian over the rollout.

| measured σ_max | reading |
|---|---|
| **≈ 1.05 – 1.20** | **striking** — an independently-derived theoretical α lands on a hand-swept empirical optimum. Strong evidence the correspondence is real and GTF is the principled version of what we built. |
| **≫ 1.2** | the formula predicts a much larger α than we measured as optimal; the correspondence is superficial and needs explaining before anything is built. |
| **< 1.0** | **GTF's α is undefined for us.** The paper states the formula requires σ_max ≥ 1. A contractive recurrent map means our exploding-gradient problem is not the one GTF solves — and per `MillerHardt2019_StableRecurrentModels`, much of sequence learning happens in the stable regime anyway. **This closes the issue.** |

**Register this threshold numerically before measuring.** It is the whole point of the issue and it costs
no GPU.

## The differences that could make the correspondence superficial

State them before testing, not after:

1. **GTF re-anchors every step; we anchor once.** Their `z̄` is the *target state at each step* — it moves.
   Our `anchor` is frozen at the end of the seed step. GTF is a continuously-refreshed pull toward truth;
   ours is a pull toward a single stale point. **These are not the same operator**, and the fact that both
   help does not make them the same mechanism.
2. **Training vs inference.** GTF shapes the *learned* dynamics; ours patches a fixed model at rollout.
   A model trained under GTF might need no inference-time blend at all.
3. **Architecture.** Their results are on shPLRNN — a compact, semi-analytically tractable RNN they
   introduce. We have a 4-cell ConvLSTM on a 180×720 grid. `σ_max` for a convolutional recurrent map over
   a spatial field is not obviously the same object.
4. **Chaos.** Their motivation is chaotic dynamics and exploding gradients. **#291 closed on exactly this
   scope condition** — conflict counts are not chaotic in the Lyapunov sense. GTF may be solving a problem
   we do not have. *(But note GTF is about bounding the Jacobian product series, which is a stability
   statement, not strictly a chaos one — that distinction is worth establishing rather than assuming.)*

## Why it is nonetheless the strongest lever currently on the board

- **It sidesteps the #288 blocker entirely.** GTF interpolates **continuous latent states**, so there is
  no count draw and no `torch.poisson` to sever the gradient. The problem that makes #288 a silent no-op
  simply does not arise.
- **We have already measured that the operator works** at inference (+0.039, paired CI excludes zero) and
  where it saturates (w≈0.1). Most training-time proposals arrive with no such prior.
- **aGTF removes the free parameter.** Claim: *"Adaptive GTF derives the forcing parameter α directly
  from the Jacobian product series of each training batch, avoiding the need for knowledge of the
  system's maximum Lyapunov exponent. An annealing strategy starts with strong forcing (α=1) that decays
  while remaining bounded from below by the batch-derived estimate."* That is a schedule, which connects
  to **#287** — and note it anneals **downward from α=1**, i.e. *increasing* teacher forcing early, which
  is the ITF direction #287 asks about.

## What to establish, in order

1. **σ_max on existing weights.** The decisive, zero-GPU check above. Register the threshold first.
2. **Is our frozen anchor a degenerate GTF, or a different operator?** If GTF's re-anchoring is what
   matters, an inference-time variant that re-anchors to a *rolling* estimate is testable cheaply on the
   existing dial harness (`run_decay_dial.sh` already takes `mode@weight`).
3. Only then: what a GTF training arm costs, and whether it subsumes the inference-time freeze.

## Falsifiers to carry

- **σ_max < 1 closes this issue.** Registered above.
- The **§4 anchor guard** (`|ΔAP(h1)| ≤ 3 × MDE`) — a training change that trades T=0 for h18 is a traded
  failure, not a retention result.
- Any comparison here is **naturally paired** (same weights, one flag) ⇒ use
  `scripts/ap_block_bootstrap.ap_diff_origin_block_ci`, **not** the between-seed MDE. See **C-306**: an
  MDE from the wrong contrast produced a verdict that was not established by the number cited.
- **M27/#291's decoupling result applies here too.** If GTF improves long-horizon behaviour, check
  whether it moved T=0 at all before crediting the mechanism.

## Literature

**In the library:**

| paper | relevance |
|---|---|
| `Hess2023_GeneralizedTeacherForcing` — *Generalized Teacher Forcing for Learning Chaotic Dynamics* | **the primary source**; the α formula, the σ_max ≥ 1 condition, aGTF's annealing |
| `MillerHardt2019_StableRecurrentModels` | if our recurrent map is stable/contractive, truncation is provably sufficient and the exploding-gradient framing does not apply — the theoretical companion to the σ_max < 1 branch |
| `Aceituno2025_TemporalHorizons` | gradient scales exponentially with horizon for chaotic systems, **linearly for limit cycles** — which regime we are in decides whether GTF's motivation transfers |
| `Teutsch2022_FlippedClassroom` (#287) | aGTF's downward annealing from α=1 *is* an increasing-TF curriculum; these two issues interact and should be sequenced, not run in parallel |
| `Vlachas2023_LearningFromPredictions` (#288) | the gradient-flow alternative that GTF sidesteps |

**To acquire:**

- **Anything applying GTF or state-space interpolation outside low-dimensional attractors** — to
  spatiotemporal grids or convolutional recurrent models. Difference (3) above is the gap.
- **Spectral-norm / Jacobian estimation for ConvLSTM** specifically, to make step 1 well-posed rather
  than improvised. `Miyato2018_SpectralNormalization` is held and is the obvious starting point for the
  power-iteration machinery, though it is used there for a different purpose.
- **Any independent replication of Hess 2023.** It is a 2023 result with a "proven" claim doing a lot of
  work here.

## Origin

Surfaced by a library sweep during the #288/#290/#291 falsifier checks (PR #293). ⚠️ Note the sweep also
found the library's search index covers **114 of 410 papers** — this paper's claims were extracted, but
several load-bearing neighbours are invisible to `library_search`, so this literature section may itself
be partial.

Related: #287, #288, #289, #280. Evidence for the w≈0.1 measurement:
`reports/2026-08-22_state_freeze_l300_dossier/` (M38/M39/M41).

**Comment by Polichinel** [comment]
## The headline falsifier cannot decide this issue — and that is worth stating plainly

This issue's central promise was that GTF **predicts our knob**, via `α = 1 − 1/σ_max`, and that σ_max is measurable on existing weights at **no GPU cost**. A threshold table was registered before measuring, which was the right thing to do.

It has been measured twice: **1.60** (retracted — wrong phase, C-308) and **7.7628** (stands, seed 43, autoregressive phase, all states converged). The previous comment computed `α = 1 − 1/7.7628 = 0.8712` against M41's `w ≈ 0.10` and **withdrew** the comparison, on grounds I want to restate because they are stronger than they look:

**σ_max is measured on a model that was never GTF-trained, while α governs one that was.** GTF changes training, so the trained model's Jacobian spectrum is not the one we can measure. The formula predicts the α for a model we do not have.

**This is not a gap in the measurement. It is a gap in the design of the falsifier**, and no amount of re-measuring closes it. The registered table cannot be applied, in either direction, and the "costs no GPU" claim was wrong — not because the measurement is expensive, but because the *decidable* version of it requires the training run the issue exists to justify.

### A second reason the summary statistic is the wrong shape

Already measured in the same run: **σ is strongly state-dependent — 3.4–7.8 early in the rollout while the state is large, ~1.47 once it has collapsed**, tracking `max|h|` almost monotonically. A single "σ_max" flattens a 5× swing. Even if the comparison were valid, one number would be the wrong summary of a quantity that varies by 5× across the horizon it is supposed to characterise.

### What would actually decide it

Not a σ measurement. Either:

1. **A cheap GTF training run** at the α its own formula prescribes, judged on the same free-running AP battery as every other intervention — accepting that this is the expensive thing the falsifier was meant to avoid; or
2. **A prior on the family.** GTF is a *training-time* intervention on the recurrent state. The training-time interventions tried here are **0-for-4**: scheduled sampling (M26–M33), ITF (M42), truncated_nb (M45), pushforward (M47). The one thing that has worked is the **inference-time** clamp (M38/M41/M48). That is not proof GTF fails, but it is the relevant base rate, and it should be stated before anyone spends a week on option 1.

### One thing this issue got right, and it has now been confirmed

The observation that we appear to have re-invented GTF's functional form at inference time still stands. More than that: the correction comment's reframing — the cell anchor works by **refilling a draining state**, not restraining a growing one — was a conjecture at the time and has since been **measured** (**M50**, emit only, ~15 min, one flag changed):

| | free-running | cell clamped |
|---|---|---|
| `max\|h\|` (seed 43) | 65.6 → 1.6, a **41× drain** | 65.6 → 66.1, **holds** |
| emitted firing rate (seed 42) | **1,547× collapse** | **2×**, flat from step 6 |

And the part that was not predictable: the **hidden** half is never clamped, yet it drains 0.98 → 0.56 free-running versus 0.98 → 0.93 clamped. Holding the cell stabilises the half it does not touch.

That reframing matters for GTF specifically. GTF's α is derived to **bound a Jacobian product series against explosion**. Our recurrence is not exploding — it is **draining**, by 41× over 36 steps, and σ falls from 7.8 to 1.47 as it does. An α derived to restrain growth is being asked to fix a decay. That is a stronger reason to doubt the correspondence than the numeric mismatch was, and it does not depend on comparing a measured σ_max to a hypothetical one.

**Suggested disposition:** keep open, but retitle away from the σ_max falsifier, and record that the decision basis is now (1) or (2) rather than a free measurement.

*(AI-agent correction, per ADR-007 — I previously described this falsifier as "unevaluated, fifteen minutes to close". It was evaluated and reasoned about more carefully than that, and I misread the thread.)*
